RecoTracker/TkSeedGenerator/src/FastCircleFit.cc

```diff
@@ -0,0 +1,148 @@
+#include "TMatrixDSymEigen.h"
+
+#ifdef MK_DEBUG
+#include <iostream>
+#define PRINT std::cout
+#else
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#define PRINT LogTrace("FastCircleFit")
+#endif
+
+namespace {
+  template<typename T>
+  T sqr(T t) { return t*t; }
+}
+
+FastCircleFit::FastCircleFit(const GlobalPoint *points, const GlobalError *errors, size_t size):
+  x0_(0), y0_(0), rho_(0),
+  chi2_(0)
+{
+  const auto N = size;
+  declareDynArray(float, N, x);
+  declareDynArray(float, N, y);
+  declareDynArray(float, N, z);
+  declareDynArray(float, N, weight); // 1/sigma^2
+
+  AlgebraicVector3 mean;
+
+  // transform
+  for(size_t i=0; i<N; ++i) {
+    const auto& point = points[i];
+    const auto p = point.basicVector();
+    x[i] = p.x();
+    y[i] = p.y();
+    z[i] = sqr(p.x()) + sqr(p.y());
+
+    // TODO: The following accounts only the hit uncertainty in phi.
+    // Albeit it "works nicely", should we account also the
+    // uncertainty in r in FPix (and the correlation between phi and r)?
+    const float phiErr2 = errors[i].phierr(point);
+    const float w = 1.f/(point.perp2()*phiErr2);
+    weight[i] = w;
+
+    PRINT << " point " << p.x()
+          << " " << p.y()
+          << " transformed " << x[i]
+          << " " << y[i]
+          << " " << z[i]
+          << " weight " << w
+          << std::endl;
+  }
+  const float invTotWeight = 1.f/std::accumulate(weight.begin(), weight.end(), 0.f);
+  PRINT << " invTotWeight " << invTotWeight;
+
+  for(size_t i=0; i<N; ++i) {
+    const float w = weight[i]*invTotWeight;
+    mean[0] += w*x[i];
+    mean[1] += w*y[i];
+    mean[2] += w*z[i];
+  }
+  PRINT << " mean " << mean[0] << " " << mean[1] << " " << mean[2] << std::endl;
+
+  // TODO: Using TMatrix for eigenvalues and eigenvectors of 3x3
+  // symmetrix matrix is an overkill. To be replaced with something
+  // lightweight.
+  TMatrixDSym A(3);
+  for(size_t i=0; i<N; ++i) {
+    const auto diff_x = x[i] - mean[0];
+    const auto diff_y = y[i] - mean[1];
+    const auto diff_z = z[i] - mean[2];
+    const auto w = weight[i];
+    A(0,0) += w * diff_x * diff_x;
+    A(0,1) += w * diff_x * diff_y;
+    A(0,2) += w * diff_x * diff_z;
+    A(1,0) += w * diff_y * diff_x;
+    A(1,1) += w * diff_y * diff_y;
+    A(1,2) += w * diff_y * diff_z;
+    A(2,0) += w * diff_z * diff_x;
+    A(2,1) += w * diff_z * diff_y;
+    A(2,2) += w * diff_z * diff_z;
+    PRINT << " i " << i << " A " << A(0,0) << " " << A(0,1) << " " << A(0,2) << std::endl
+          << "     " << A(1,0) << " " << A(1,1) << " " << A(1,2) << std::endl
+          << "     " << A(2,0) << " " << A(2,1) << " " << A(2,2) << std::endl;
+  }
+  A *= 1./static_cast<double>(N);
+
+  PRINT << " A " << A(0,0) << " " << A(0,1) << " " << A(0,2) << std::endl
+        << "   " << A(1,0) << " " << A(1,1) << " " << A(1,2) << std::endl
+        << "   " << A(2,0) << " " << A(2,1) << " " << A(2,2) << std::endl;
+
+  // find eigenvalues and vectors
```

Comment:
@slava77 Ah, sorry, I'll to follow up later (and will definitively try eigen, the interface looked even simpler than TMatrix). TMatrix is good-enough to prove the point (anyway the circle fit is only ~2 % of the quadruplet seeding time).